client/app/pages/groups/GroupDataSources.jsx

```diff
@@ -0,0 +1,262 @@
+    },
+  ];
+
+  listColumns = [
+    Columns.custom((text, datasource) => (
+      <DataSourcePreviewCard dataSource={datasource} withLink />
+    ), {
+      title: 'Name',
+      field: 'name',
+      width: null,
+    }),
+    Columns.custom((text, datasource) => {
+      const menu = (
+        <Menu
+          selectedKeys={[datasource.view_only ? 'viewonly' : 'full']}
+          onClick={item => this.setDataSourcePermissions(item.domEvent, datasource, item.key)}
+        >
+          <Menu.Item key="full">Full Access</Menu.Item>
+          <Menu.Item key="viewonly">View Only</Menu.Item>
+        </Menu>
+      );
+
+      const cancelEvent = (event) => {
+        event.preventDefault();
+        event.stopPropagation();
+      };
+
+      return (
+        <Dropdown trigger={['click']} overlay={menu}>
+          <Button className="w-100" onClick={cancelEvent}>
+            {datasource.view_only ? 'View Only' : 'Full Access'} <Icon type="down" />
+          </Button>
+        </Dropdown>
+      );
+    }, {
+      width: '1%',
+      className: 'p-r-0',
+      isAvailable: () => currentUser.isAdmin,
+    }),
+    Columns.custom((text, datasource) => (
+      <Button className="w-100" type="danger" onClick={event => this.removeGroupDataSource(event, datasource)}>Remove</Button>
+    ), {
+      width: '1%',
+      isAvailable: () => currentUser.isAdmin,
+    }),
+  ];
+
+  componentDidMount() {
+    Group.get({ id: this.groupId }).$promise.then((group) => {
+      this.group = group;
+      this.forceUpdate();
+    });
+  }
+
+  onTableRowClick = (event, item) => navigateTo('data_sources/' + item.id);
+
+  removeGroupDataSource = (event, datasource) => {
+    // prevent default click action on table rows
+    event.preventDefault();
+    event.stopPropagation();
+
+    $http.delete(`api/groups/${this.group.id}/data_sources/${datasource.id}`).success(() => {
+      this.props.controller.update();
+    });
+  };
+
+  setDataSourcePermissions = (event, datasource, permission) => {
+    // prevent default click action on table rows
+    event.preventDefault();
+    event.stopPropagation();
+
+    const viewOnly = permission !== 'full';
+
+    $http.post(`api/groups/${this.group.id}/data_sources/${datasource.id}`, { view_only: viewOnly }).success(() => {
+      datasource.view_only = viewOnly;
+      this.forceUpdate();
+    });
+  };
+
+  addDataSources = () => {
+    const allDataSources = DataSource.query({ extended: true }).$promise;
+    SelectItemsDialog.showModal({
+      dialogTitle: 'Add Data Sources',
+      inputPlaceholder: 'Search data sources...',
+      selectedItemsTitle: 'New Data Sources',
+      searchItems: (searchTerm) => {
+        searchTerm = searchTerm.toLowerCase();
+        return allDataSources.then(items => filter(items, ds => ds.name.toLowerCase().includes(searchTerm)));
+      },
+      renderItem: (item, { isSelected }) => {
+        const alreadyInGroup = isBoolean((item.groups || {})[this.groupId]);
```

Comment:
This might have a bug. Already selected data sources are still selectable (results in Internal Server Error).

Reply:
That's what I mentioned earlier - https://github.com/getredash/redash/pull/3411#issuecomment-465630332 (6)